src/redis-cli.c

```diff
@@ -7726,40 +7739,62 @@ unsigned long long sendSync(redisContext *c, char *out_eof) {
         fprintf(stderr, "SYNC with master failed: %s\n", buf);
         exit(1);
     }
+
+    /* Handling PSYNC responses. */
+    if (!strncmp(buf, "+FULLRESYNC", 11) ||
+        !strncmp(buf, "+CONTINUE", 9))
+    {
+        int sync_partial = !strncmp(buf, "+CONTINUE", 9);
+        fprintf(stderr, "PSYNC replied %s\n", buf);
+        p = buf;
+        while(1) {
+            nread = readConn(c,p,1);
+            if (nread <= 0) {
+                fprintf(stderr,"Error reading bulk length while PSYNCing\n");
+                exit(1);
+            }
+            if (*p == '\n' && p != buf) break;
+            if (*p != '\n') p++;
+        }
+        *p = '\0';
+
+        if (sync_partial) return 0;
+    }
+
     if (strncmp(buf+1,"EOF:",4) == 0 && strlen(buf+5) >= RDB_EOF_MARK_SIZE) {
         memcpy(out_eof, buf+5, RDB_EOF_MARK_SIZE);
         return 0;
     }
     return strtoull(buf+1,NULL,10);
 }
 
-static void slaveMode(void) {
+static void slaveMode(int is_psync) {
     static char eofmark[RDB_EOF_MARK_SIZE];
     static char lastbytes[RDB_EOF_MARK_SIZE];
     static int usemark = 0;
-    unsigned long long payload = sendSync(context,eofmark);
+    unsigned long long payload = sendSync(context,eofmark,!is_psync);
     char buf[1024];
     int original_output = config.output;
+    char *command = is_psync ? "PSYNC" : "SYNC";
 
-    if (payload == 0) {
+    if (payload == 0 && !is_psync) {
```

Comment:
since `sendSync` returns the same (0) on both CONTINUE and EOF, i don't understand how we can tell them apart.
the type of SYNC/PSYNC command used is unrelated to this.

Reply:
it's a problem, do you have any ideas... ?

Reply:
i think the return value of `sendSync` should be an indication of the partial / full mode, and we should have an additional output parameter indicating the size of the payload in case of full mode,
if that payload output value is 0, and we're in full mode, it's an indication that we should use the EOF marker.

Reply:
> but with CONTINUE isn't the next thing we get is the command stream?

that is true, the while 1 is actually needed, there are a few `\n` we need to read.

> btw, if the user declared REPLCONF CAPA PSYNC2, we may need to parse a more complicated +CONTINUE.
your code seems to already handle that, right?

yes, it can handle that